tests/unittests/dhcp/kea_metrics_test.py

```diff
@@ -0,0 +1,698 @@
+    assert set(source.fetch_metrics()) == set(expected_metrics)
+
+
+def test_dhcp4_config_and_statistic_response_that_is_valid_should_return_every_metric(
+    valid_dhcp4, responsequeue
+):
+    config, statistics, expected_metrics = valid_dhcp4
+    responsequeue.prefill("dhcp4", config, statistics)
+    source = KeaDhcpMetricSource("192.0.1.2", 80, dhcp_version=4, tzinfo=timezone.utc)
+    assert set(source.fetch_metrics()) == set(expected_metrics)
+
+
+@pytest.mark.parametrize(
+    "status", [status for status in KeaStatus if status != KeaStatus.SUCCESS]
+)
+def test_config_response_with_error_status_should_raise_KeaError(
+    valid_dhcp4, responsequeue, status
+):
+    """
+    If Kea responds with an error while fetching the Kea DHCP's config during
+    fetch_metrics(), we cannot continue further, so we fail.
+    """
+    config, statistics, _ = valid_dhcp4
+    responsequeue.prefill("dhcp4", None, statistics)
+    responsequeue.add("config-get", kearesponse(config, status=status))
+    source = KeaDhcpMetricSource("192.0.1.2", 80, dhcp_version=4)
+    with pytest.raises(KeaError):
+        source.fetch_metrics()
+
+
+def test_any_response_with_invalid_format_should_raise_KeaError(
+    valid_dhcp4, responsequeue
+):
+    """
+    If Kea responds with an invalid format (i.e. in an unrecognizable way), we
+    should fail loudly, because chances are either the host we're sending
+    requests to is not a Kea Control Agent, or there's a part of the API that
+    we've not covered.
+    """
+    config, statistics, _ = valid_dhcp4
+    source = KeaDhcpMetricSource("192.0.1.2", 80, dhcp_version=4)
+
+    responsequeue.add("config-get", "{}")
+    with pytest.raises(KeaError):
+        source.fetch_metrics()
+
+    responsequeue.clear()
+
+    responsequeue.prefill("dhcp4", config, None)
+    responsequeue.add("statistic-get", "{}")
+    with pytest.raises(KeaError):
+        source.fetch_metrics()
+
+    responsequeue.clear()
+
+    responsequeue.prefill("dhcp4", None, statistics)
+    responsequeue.add("config-get", "{}")
+    with pytest.raises(KeaError):
+        source.fetch_metrics()
+
+    responsequeue.clear()
+
+    # config-hash-get is only called if some config-get includes a hash we can compare
+    # with the next time we're attempting to fetch a config:
+    config["Dhcp4"]["hash"] = "foo"
+    responsequeue.prefill("dhcp4", config, statistics)
+    responsequeue.add("config-hash-get", "{}")
+    with pytest.raises(KeaError):
+        source.fetch_metrics()
+
+
+def test_all_responses_is_empty_but_valid_should_yield_no_metrics(
+    valid_dhcp4, responsequeue
+):
+    """
+    If the Kea DHCP server we query does not have any subnets configured, the
+    correct thing to do is to return an empty iterable, (as opposed to failing).
+
+    Likewise, if it returns no statistics for its configured subnets, the
+    correct thing to do is to return an empty iterable.
+    """
```

Comment:
Can this be split into 2 tests? One for "no subnets configured" and one "no statistics for configured subnets"